### ISSUE TITLE
Logging: Need to add secure logging APIs to OE SDK

### ISSUE THREAD

**Issue by soccerGB** [openenclave/openenclave#1046]
Currently, this is no secure logging API for OE application.  Debugging an OE relies on the availability of the access to the system. A secure API would greatly help the debugging an OE issues on a remote machine.

**Comment by soccerGB** [comment]
.      Logging: Need to add secure logging APIs to OE SDK
 
Currently, our OE SDK  does not have any “secure logging API” for OE applications to log enclave specific information.  Adding “secure logging API”  to OE SDK would help OE enclave developers triage enclave issues reported from their enclave users. 
I chatted with some people on the team (thanks a lot for the input from Istvan). It turns out there are more difficulties than that I expected when it comes to  implementing a secure logging API for enclaves, mainly because the host is untrusted and assumed to be malicious.
Here are the key considering points:
a.	Data ownership: 
a.	Should the owner of the logs be the same as the owner of the data logged?
b.	Can we assume the owner of the data same as the developer of the enclave?
c.	An individual enclave instance can include data from multiple entities. This is called secure multiparty computation. Who owns the log for such an enclave?
c) There is no client authentication built into the enclave SDK. But accepting/releasing session keys associated with logging must go through some kind of client authentication as per point "a".
 
I don’t have all the answers for above questions.  It might take some brainstorming exercise to get a reasonable solution. We are welcome to share your thoughts.

**Comment by soccerGB** [comment]
With the filesystem support, could we consider adding support for this?

**Comment by radhikaj** [comment]
We have no requirements for this as yet.

**Comment by soccerGB** [comment]
The existing OE logging is only for debugging purpose (can only be enabled in debugged binary)
Need to have a logging story for production solution, which was why this issue was opened.

**Comment by jhand2** [comment]
I'm not convinced this is something that OE should support directly since it's pretty dependent on the enclave application.

For example, does this mean spooling logs into an encrypted file on the host? In that case the requirement is really support for an encrypted filesystem.

Does this mean streaming logs to some https service? Then really https support is what you want.

I think the requirements are so custom that asking customers to write their own logging libraries that meet their needs makes more sense. We should just give them the tools to write arbitrary data security to arbitrary locations.

**Comment by achamayou** [comment]
@jhand2 I agree, I don't think OE itself ought to provide this, there can be a library (openenclave/logging) that's offered to applications that require it, but I don't think it belongs in the core SDK.

**Comment by soccerGB** [comment]
Is the suggestion here to ask OE application developers to build their own logging library for their own consumption?

**Comment by jhand2** [comment]
I'm open to suggestions but logging needs are very application specific and I don't think it makes sense for OE to provide extra logging functionality. I asked a bunch of questions about the use case for this issue and didn't get any answers.

If a user of the SDK has a specific need for this type of feature, that might help make this issue more concrete.

**Comment by jhand2** [comment]
Requesting to close this issue as a dup of #2895 since I think that issue is more concrete ask.

**Comment by soccerGB** [comment]
This is for secure logging that enables an application to log their own logging securely,  which is totally different from #2985, which requests a way to redirect existing Oe SDK logging events to custom destination.

**Comment by radhikaj** [comment]
This is a duplicate of #3075
